common/solver/idr_kernels.hpp.inc

```diff
@@ -79,6 +80,8 @@ __global__
             __syncthreads();
 
             dot = reduction_helper[0];
+            // avoid the first thread writing it before every threads read it.
+            __syncthreads();
```

Comment:
I think we can remove this and move the first __syncthreads() before `reduction_helper[tidx] = dot`, because `reduce` already synchronizes before accessing memory. Not sure whether the CUDA compiler removes duplicate barrier.sync instructions.

Reply:
I guess you mean removing the second one?
I will need to add another one before `reduction_helper_real[tidx] = norm;` to avoid `reduction_helper_real[tidx]` issue because they are on the same memory.

Reply:
I mean
```cpp
for (...) {
__syncthreads();
reduction_helper[tidx] = dot;
reduce(...);
__syncthreads();
dot = reduction_helper[0];
}
...
__syncthreads();
reduction_helper_real[tidx] = norm;
reduce(...);
__syncthreads();
norm = reduction_helper_real[0];
```

Reply:
I see. I am fine with both version